File: benchmarks/java/DataStructures.java
/**
 * Ruva Data Structures Library — Java Reference Implementations
 * 
 * This file contains core data structure implementations that mirror
 * what Ruva's stdlib provides. Used for:
 *   1. Cross-language benchmarking
 *   2. Reference implementation for Ruva transpiler output
 *   3. Algorithm testing
 * 
 * Structures: ArrayList, LinkedList, HashMap, Stack, Queue, BinaryTree, Heap
 */

import java.util.*;
import java.util.function.*;

public class DataStructures {

    // ═══════════════════════════════════════════════════════════════════════════
    // ArrayList — Dynamic array with automatic resizing
    // ═══════════════════════════════════════════════════════════════════════════

    public static class ArrayList<T> {
        private Object[] data;
        private int size;
        private static final int DEFAULT_CAPACITY = 16;

        public ArrayList() {
            this.data = new Object[DEFAULT_CAPACITY];
            this.size = 0;
        }

        public ArrayList(int initialCapacity) {
            this.data = new Object[Math.max(initialCapacity, 1)];
            this.size = 0;
        }

        public void add(T item) {
            ensureCapacity(size + 1);
            data[size++] = item;
        }

        public void add(int index, T item) {
            rangeCheckForAdd(index);
            ensureCapacity(size + 1);
            System.arraycopy(data, index, data, index + 1, size - index);
            data[index] = item;
            size++;
        }

        @SuppressWarnings("unchecked")
        public T get(int index) {
            rangeCheck(index);
            return (T) data[index];
        }

        @SuppressWarnings("unchecked")
        public T set(int index, T item) {
            rangeCheck(index);
            T old = (T) data[index];
            data[index] = item;
            return old;
        }

        @SuppressWarnings("unchecked")
        public T remove(int index) {
            rangeCheck(index);
            T old = (T) data[index];
            int numMoved = size - index - 1;
            if (numMoved > 0) {
                System.arraycopy(data, index + 1, data, index, numMoved);
            }
            data[--size] = null;
            return old;
        }

        public boolean remove(T item) {
            for (int i = 0; i < size; i++) {
                if (Objects.equals(data[i], item)) {
                    remove(i);
                    return true;
                }
            }
            return false;
        }

        public int size() { return size; }
        public boolean isEmpty() { return size == 0; }

        public void clear() {
            for (int i = 0; i < size; i++) data[i] = null;
            size = 0;
        }

        private void ensureCapacity(int minCapacity) {
            if (minCapacity > data.length) {
                int newCapacity = data.length + (data.length >> 1);
                data = Arrays.copyOf(data, Math.max(newCapacity, minCapacity));
            }
        }

        private void rangeCheck(int index) {
            if (index < 0 || index >= size)
                throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
        }

        private void rangeCheckForAdd(int index) {
            if (index < 0 || index > size)
                throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
        }
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // LinkedList — Doubly-linked list
    // ═══════════════════════════════════════════════════════════════════════════

    public static class LinkedList<T> {
        private static class Node<T> {
            T data;
            Node<T> next;
            Node<T> prev;

            Node(T data) {
                this.data = data;
            }
        }

        private Node<T> head;
        private Node<T> tail;
        private int size;

        public LinkedList() {
            head = null;
            tail = null;
            size = 0;
        }

        public void addFirst(T item) {
            Node<T> node = new Node<>(item);
            if (head == null) {
                head = tail = node;
            } else {
                node.next = head;
                head.prev = node;
                head = node;
            }
            size++;
        }

        public void addLast(T item) {
            Node<T> node = new Node<>(item);
            if (tail == null) {
                head = tail = node;
            } else {
                tail.next = node;
                node.prev = tail;
                tail = node;
            }
            size++;
        }

        public T removeFirst() {
            if (head == null) throw new NoSuchElementException();
            T data = head.data;
            head = head.next;
            if (head == null) tail = null;
            else head.prev = null;
            size--;
            return data;
        }

        public T removeLast() {
            if (tail == null) throw new NoSuchElementException();
            T data = tail.data;
            tail = tail.prev;
            if (tail == null) head = null;
            else tail.next = null;
            size--;
            return data;
        }

        public T getFirst() {
            if (head == null) throw new NoSuchElementException();
            return head.data;
        }

        public T getLast() {
            if (tail == null) throw new NoSuchElementException();
            return tail.data;
        }

        public int size() { return size; }
        public boolean isEmpty() { return size == 0; }

        public boolean contains(T item) {
            Node<T> current = head;
            while (current != null) {
                if (Objects.equals(current.data, item)) return true;
                current = current.next;
            }
            return false;
        }

        public void forEach(Consumer<T> action) {
            Node<T> current = head;
            while (current != null) {
                action.accept(current.data);
                current = current.next;
            }
        }
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // HashMap — Hash table with separate chaining
    // ═══════════════════════════════════════════════════════════════════════════

    public static class HashMap<K, V> {
        private static class Entry<K, V> {
            K key;
            V value;
            Entry<K, V> next;
            int hash;

            Entry(K key, V value, int hash) {
                this.key = key;
                this.value = value;
                this.hash = hash;
            }
        }

        private Entry<K, V>[] table;
        private int size;
        private static final int DEFAULT_CAPACITY = 16;
        private static final float DEFAULT_LOAD_FACTOR = 0.75f;

        @SuppressWarnings("unchecked")
        public HashMap() {
            table = new Entry[DEFAULT_CAPACITY];
            size = 0;
        }

        public V put(K key, V value) {
            int hash = key.hashCode();
            int index = indexFor(hash, table.length);

            Entry<K, V> current = table[index];
            while (current != null) {
                if (current.hash == hash && Objects.equals(current.key, key)) {
                    V old = current.value;
                    current.value = value;
                    return old;
                }
                current = current.next;
            }

            Entry<K, V> newEntry = new Entry<>(key, value, hash);
            newEntry.next = table[index];
            table[index] = newEntry;
            size++;

            if (size > table.length * DEFAULT_LOAD_FACTOR) {
                resize();
            }
            return null;
        }

        public V get(K key) {
            int hash = key.hashCode();
            int index = indexFor(hash, table.length);

            Entry<K, V> current = table[index];
            while (current != null) {
                if (current.hash == hash && Objects.equals(current.key, key)) {
                    return current.value;
                }
                current = current.next;
            }
            return null;
        }

        public V remove(K key) {
            int hash = key.hashCode();
            int index = indexFor(hash, table.length);

            Entry<K, V> prev = null;
            Entry<K, V> current = table[index];
            while (current != null) {
                if (current.hash == hash && Objects.equals(current.key, key)) {
                    if (prev == null) table[index] = current.next;
                    else prev.next = current.next;
                    size--;
                    return current.value;
                }
                prev = current;
                current = current.next;
            }
            return null;
        }

        public boolean containsKey(K key) {
            return get(key) != null;
        }

        public int size() { return size; }
        public boolean isEmpty() { return size == 0; }

        @SuppressWarnings("unchecked")
        private void resize() {
            Entry<K, V>[] oldTable = table;
            table = new Entry[oldTable.length * 2];
            size = 0;
            for (Entry<K, V> entry : oldTable) {
                Entry<K, V> current = entry;
                while (current != null) {
                    put(current.key, current.value);
                    current = current.next;
                }
            }
        }

        private int indexFor(int hash, int length) {
            return Math.abs(hash) % length;
        }
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // Stack — LIFO data structure
    // ═══════════════════════════════════════════════════════════════════════════

    public static class Stack<T> {
        private final LinkedList<T> list = new LinkedList<>();

        public void push(T item) {
            list.addFirst(item);
        }

        public T pop() {
            return list.removeFirst();
        }

        public T peek() {
            return list.getFirst();
        }

        public int size() { return list.size(); }
        public boolean isEmpty() { return list.isEmpty(); }
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // Queue — FIFO data structure
    // ═══════════════════════════════════════════════════════════════════════════

    public static class Queue<T> {
        private final LinkedList<T> list = new LinkedList<>();

        public void enqueue(T item) {
            list.addLast(item);
        }

        public T dequeue() {
            return list.removeFirst();
        }

        public T peek() {
            return list.getFirst();
        }

        public int size() { return list.size(); }
        public boolean isEmpty() { return list.isEmpty(); }
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // BinaryTree — Binary search tree
    // ═══════════════════════════════════════════════════════════════════════════

    public static class BinaryTree<T extends Comparable<T>> {
        private static class TreeNode<T> {
            T data;
            TreeNode<T> left, right;

            TreeNode(T data) {
                this.data = data;
            }
        }

        private TreeNode<T> root;
        private int size;

        public BinaryTree() {
            root = null;
            size = 0;
        }

        public void insert(T item) {
            root = insertRec(root, item);
            size++;
        }

        private TreeNode<T> insertRec(TreeNode<T> node, T item) {
            if (node == null) return new TreeNode<>(item);
            if (item.compareTo(node.data) < 0) {
                node.left = insertRec(node.left, item);
            } else if (item.compareTo(node.data) > 0) {
                node.right = insertRec(node.right, item);
            }
            return node;
        }

        public boolean contains(T item) {
            return searchRec(root, item);
        }

        private boolean searchRec(TreeNode<T> node, T item) {
            if (node == null) return false;
            if (item.compareTo(node.data) == 0) return true;
            if (item.compareTo(node.data) < 0) return searchRec(node.left, item);
            return searchRec(node.right, item);
        }

        public List<T> inorder() {
            List<T> result = new ArrayList<>();
            inorderRec(root, result);
            return result;
        }

        private void inorderRec(TreeNode<T> node, List<T> result) {
            if (node == null) return;
            inorderRec(node.left, result);
            result.add(node.data);
            inorderRec(node.right, result);
        }

        public int size() { return size; }

        public int height() {
            return heightRec(root);
        }

        private int heightRec(TreeNode<T> node) {
            if (node == null) return 0;
            return 1 + Math.max(heightRec(node.left), heightRec(node.right));
        }
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // MinHeap — Priority queue implementation
    // ═══════════════════════════════════════════════════════════════════════════

    public static class MinHeap<T extends Comparable<T>> {
        private ArrayList<T> heap;

        public MinHeap() {
            heap = new ArrayList<>();
        }

        public void add(T item) {
            heap.add(item);
            siftUp(heap.size() - 1);
        }

        public T poll() {
            if (heap.isEmpty()) throw new NoSuchElementException();
            T min = heap.get(0);
            T last = heap.remove(heap.size() - 1);
            if (!heap.isEmpty()) {
                heap.set(0, last);
                siftDown(0);
            }
            return min;
        }

        public T peek() {
            if (heap.isEmpty()) throw new NoSuchElementException();
            return heap.get(0);
        }

        public int size() { return heap.size(); }
        public boolean isEmpty() { return heap.isEmpty(); }

        private void siftUp(int index) {
            while (index > 0) {
                int parent = (index - 1) / 2;
                if (heap.get(index).compareTo(heap.get(parent)) < 0) {
                    swap(index, parent);
                    index = parent;
                } else break;
            }
        }

        private void siftDown(int index) {
            int size = heap.size();
            while (true) {
                int left = 2 * index + 1;
                int right = 2 * index + 2;
                int smallest = index;

                if (left < size && heap.get(left).compareTo(heap.get(smallest)) < 0) {
                    smallest = left;
                }
                if (right < size && heap.get(right).compareTo(heap.get(smallest)) < 0) {
                    smallest = right;
                }
                if (smallest != index) {
                    swap(index, smallest);
                    index = smallest;
                } else break;
            }
        }

        private void swap(int i, int j) {
            T temp = heap.get(i);
            heap.set(i, heap.get(j));
            heap.set(j, temp);
        }
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // Union-Find — Disjoint set data structure
    // ═══════════════════════════════════════════════════════════════════════════

    public static class UnionFind {
        private int[] parent;
        private int[] rank;
        private int components;

        public UnionFind(int n) {
            parent = new int[n];
            rank = new int[n];
            components = n;
            for (int i = 0; i < n; i++) {
                parent[i] = i;
                rank[i] = 0;
            }
        }

        public int find(int x) {
            if (parent[x] != x) {
                parent[x] = find(parent[x]); // Path compression
            }
            return parent[x];
        }

        public boolean union(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);
            if (rootX == rootY) return false;

            // Union by rank
            if (rank[rootX] < rank[rootY]) {
                parent[rootX] = rootY;
            } else if (rank[rootX] > rank[rootY]) {
                parent[rootY] = rootX;
            } else {
                parent[rootY] = rootX;
                rank[rootX]++;
            }
            components--;
            return true;
        }

        public boolean connected(int x, int y) {
            return find(x) == find(y);
        }

        public int components() { return components; }
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // Demo / Test
    // ═══════════════════════════════════════════════════════════════════════════

    public static void main(String[] args) {
        System.out.println("=== Ruva Data Structures Library (Java) ===\n");

        // ArrayList test
        ArrayList<String> list = new ArrayList<>();
        list.add("alpha");
        list.add("beta");
        list.add("gamma");
        System.out.println("ArrayList: " + list.get(0) + ", " + list.get(1) + ", " + list.get(2));
        System.out.println("  size: " + list.size());

        // Stack test
        Stack<Integer> stack = new Stack<>();
        stack.push(10);
        stack.push(20);
        stack.push(30);
        System.out.println("\nStack pop: " + stack.pop() + ", " + stack.pop());

        // HashMap test
        HashMap<String, Integer> map = new HashMap<>();
        map.put("one", 1);
        map.put("two", 2);
        map.put("three", 3);
        System.out.println("\nHashMap: one=" + map.get("one") + ", two=" + map.get("two"));

        // BinaryTree test
        BinaryTree<Integer> tree = new BinaryTree<>();
        for (int i : new int[]{5, 3, 7, 1, 4, 6, 8}) {
            tree.insert(i);
        }
        System.out.println("\nBST inorder: " + tree.inorder());
        System.out.println("  height: " + tree.height());

        // MinHeap test
        MinHeap<Integer> heap = new MinHeap<>();
        for (int i : new int[]{5, 3, 7, 1, 4, 6, 8}) {
            heap.add(i);
        }
        System.out.print("\nMinHeap poll order: ");
        while (!heap.isEmpty()) {
            System.out.print(heap.poll() + " ");
        }
        System.out.println();

        // Union-Find test
        UnionFind uf = new UnionFind(6);
        uf.union(0, 1);
        uf.union(1, 2);
        uf.union(3, 4);
        System.out.println("\nUnionFind: 0~2 connected? " + uf.connected(0, 2));
        System.out.println("  0~3 connected? " + uf.connected(0, 3));
        System.out.println("  components: " + uf.components());

        System.out.println("\nAll data structures functional.");
    }
}

File: benchmarks/python/data_structures.py
#!/usr/bin/env python3
"""
Ruva Data Structures Library — Python Reference Implementations

Core data structures for benchmarking and reference:
  - ArrayList (dynamic array with type hints)
  - HashMap (dictionary with ordered iteration)
  - Stack (LIFO)
  - Queue (FIFO with deque)
  - BinarySearchTree
  - MinHeap
  - UnionFind
  - Trie (prefix tree)
  - LRU Cache
"""

from __future__ import annotations
from typing import Any, Callable, Generic, Iterable, Iterator, List, Optional, Tuple, TypeVar
from collections import deque
import time

T = TypeVar("T")
K = TypeVar("K")
V = TypeVar("V")


# ═══════════════════════════════════════════════════════════════════════════
# ArrayList — Dynamic array with extra utilities
# ═══════════════════════════════════════════════════════════════════════════

class ArrayList(Generic[T]):
    """Dynamic array — like Python's list but with explicit capacity management."""

    def __init__(self, capacity: int = 16) -> None:
        self._data: list[T] = []
        self._capacity = capacity

    def add(self, item: T) -> None:
        self._data.append(item)

    def add_at(self, index: int, item: T) -> None:
        if index < 0 or index > len(self._data):
            raise IndexError(f"Index: {index}, Size: {len(self._data)}")
        self._data.insert(index, item)

    def get(self, index: int) -> T:
        if index < 0 or index >= len(self._data):
            raise IndexError(f"Index: {index}, Size: {len(self._data)}")
        return self._data[index]

    def set(self, index: int, item: T) -> T:
        if index < 0 or index >= len(self._data):
            raise IndexError(f"Index: {index}, Size: {len(self._data)}")
        old = self._data[index]
        self._data[index] = item
        return old

    def remove_at(self, index: int) -> T:
        if index < 0 or index >= len(self._data):
            raise IndexError(f"Index: {index}, Size: {len(self._data)}")
        return self._data.pop(index)

    def remove(self, item: T) -> bool:
        try:
            self._data.remove(item)
            return True
        except ValueError:
            return False

    def contains(self, item: T) -> bool:
        return item in self._data

    def index_of(self, item: T) -> int:
        try:
            return self._data.index(item)
        except ValueError:
            return -1

    def size(self) -> int:
        return len(self._data)

    def is_empty(self) -> bool:
        return len(self._data) == 0

    def clear(self) -> None:
        self._data.clear()

    def to_list(self) -> list[T]:
        return list(self._data)

    def __repr__(self) -> str:
        return f"ArrayList({self._data})"


# ═══════════════════════════════════════════════════════════════════════════
# HashMap — Hash table with chaining
# ═══════════════════════════════════════════════════════════════════════════

class HashMap(Generic[K, V]):
    """Hash map with separate chaining for collision resolution."""

    def __init__(self, capacity: int = 16, load_factor: float = 0.75) -> None:
        self._capacity = capacity
        self._load_factor = load_factor
        self._buckets: list[list[Tuple[K, V]]] = [[] for _ in range(capacity)]
        self._size = 0

    def _hash(self, key: K) -> int:
        return hash(key) % self._capacity

    def put(self, key: K, value: V) -> Optional[V]:
        if self._size >= self._capacity * self._load_factor:
            self._resize()

        bucket_idx = self._hash(key)
        bucket = self._buckets[bucket_idx]

        for i, (k, v) in enumerate(bucket):
            if k == key:
                old = v
                bucket[i] = (key, value)
                return old

        bucket.append((key, value))
        self._size += 1
        return None

    def get(self, key: K) -> Optional[V]:
        bucket_idx = self._hash(key)
        for k, v in self._buckets[bucket_idx]:
            if k == key:
                return v
        return None

    def remove(self, key: K) -> Optional[V]:
        bucket_idx = self._hash(key)
        bucket = self._buckets[bucket_idx]
        for i, (k, v) in enumerate(bucket):
            if k == key:
                bucket.pop(i)
                self._size -= 1
                return v
        return None

    def contains_key(self, key: K) -> bool:
        return self.get(key) is not None

    def size(self) -> int:
        return self._size

    def is_empty(self) -> bool:
        return self._size == 0

    def keys(self) -> list[K]:
        result = []
        for bucket in self._buckets:
            for k, _ in bucket:
                result.append(k)
        return result

    def values(self) -> list[V]:
        result = []
        for bucket in self._buckets:
            for _, v in bucket:
                result.append(v)
        return result

    def items(self) -> list[Tuple[K, V]]:
        result = []
        for bucket in self._buckets:
            result.extend(bucket)
        return result

    def _resize(self) -> None:
        old_items = self.items()
        self._capacity *= 2
        self._buckets = [[] for _ in range(self._capacity)]
        self._size = 0
        for k, v in old_items:
            self.put(k, v)

    def __repr__(self) -> str:
        return f"HashMap({dict(self.items())})"


# ═══════════════════════════════════════════════════════════════════════════
# Stack — LIFO data structure
# ═══════════════════════════════════════════════════════════════════════════

class Stack(Generic[T]):
    """LIFO stack — push, pop, peek."""

    def __init__(self) -> None:
        self._items: list[T] = []

    def push(self, item: T) -> None:
        self._items.append(item)

    def pop(self) -> T:
        if not self._items:
            raise IndexError("Stack is empty")
        return self._items.pop()

    def peek(self) -> T:
        if not self._items:
            raise IndexError("Stack is empty")
        return self._items[-1]

    def size(self) -> int:
        return len(self._items)

    def is_empty(self) -> bool:
        return len(self._items) == 0

    def clear(self) -> None:
        self._items.clear()

    def __repr__(self) -> str:
        return f"Stack({self._items})"


# ═══════════════════════════════════════════════════════════════════════════
# Queue — FIFO data structure (deque-based)
# ═══════════════════════════════════════════════════════════════════════════

class Queue(Generic[T]):
    """FIFO queue — enqueue, dequeue, peek."""

    def __init__(self) -> None:
        self._items: deque[T] = deque()

    def enqueue(self, item: T) -> None:
        self._items.append(item)

    def dequeue(self) -> T:
        if not self._items:
            raise IndexError("Queue is empty")
        return self._items.popleft()

    def peek(self) -> T:
        if not self._items:
            raise IndexError("Queue is empty")
        return self._items[0]

    def size(self) -> int:
        return len(self._items)

    def is_empty(self) -> bool:
        return len(self._items) == 0

    def __repr__(self) -> str:
        return f"Queue({list(self._items)})"


# ═══════════════════════════════════════════════════════════════════════════
# BinarySearchTree — Ordered tree
# ═══════════════════════════════════════════════════════════════════════════

class BinarySearchTree(Generic[T]):
    """Binary search tree with insert, search, and traversal."""

    class _Node:
        def __init__(self, data: T) -> None:
            self.data = data
            self.left: Optional["BinarySearchTree._Node"] = None
            self.right: Optional["BinarySearchTree._Node"] = None

    def __init__(self) -> None:
        self.root: Optional[BinarySearchTree._Node] = None
        self.size = 0

    def insert(self, data: T) -> None:
        self.root = self._insert_rec(self.root, data)
        self.size += 1

    def _insert_rec(self, node: Optional[BinarySearchTree._Node], data: T) -> BinarySearchTree._Node:
        if node is None:
            return BinarySearchTree._Node(data)
        if data < node.data:
            node.left = self._insert_rec(node.left, data)
        elif data > node.data:
            node.right = self._insert_rec(node.right, data)
        return node

    def contains(self, data: T) -> bool:
        return self._search_rec(self.root, data)

    def _search_rec(self, node: Optional[BinarySearchTree._Node], data: T) -> bool:
        if node is None:
            return False
        if data == node.data:
            return True
        if data < node.data:
            return self._search_rec(node.left, data)
        return self._search_rec(node.right, data)

    def inorder(self) -> list[T]:
        result: list[T] = []
        self._inorder_rec(self.root, result)
        return result

    def _inorder_rec(self, node: Optional[BinarySearchTree._Node], result: list[T]) -> None:
        if node is None:
            return
        self._inorder_rec(node.left, result)
        result.append(node.data)
        self._inorder_rec(node.right, result)

    def height(self) -> int:
        return self._height_rec(self.root)

    def _height_rec(self, node: Optional[BinarySearchTree._Node]) -> int:
        if node is None:
            return 0
        return 1 + max(self._height_rec(node.left), self._height_rec(node.right))

    def __repr__(self) -> str:
        return f"BST({self.inorder()})"


# ═══════════════════════════════════════════════════════════════════════════
# MinHeap — Priority queue
# ═══════════════════════════════════════════════════════════════════════════

class MinHeap(Generic[T]):
    """Min-heap for priority queue operations."""

    def __init__(self) -> None:
        self._data: list[T] = []

    def push(self, item: T) -> None:
        self._data.append(item)
        self._sift_up(len(self._data) - 1)

    def pop(self) -> T:
        if not self._data:
            raise IndexError("Heap is empty")
        min_val = self._data[0]
        last = self._data.pop()
        if self._data:
            self._data[0] = last
            self._sift_down(0)
        return min_val

    def peek(self) -> T:
        if not self._data:
            raise IndexError("Heap is empty")
        return self._data[0]

    def size(self) -> int:
        return len(self._data)

    def is_empty(self) -> bool:
        return len(self._data) == 0

    def _sift_up(self, index: int) -> None:
        while index > 0:
            parent = (index - 1) // 2
            if self._data[index] < self._data[parent]:
                self._data[index], self._data[parent] = self._data[parent], self._data[index]
                index = parent
            else:
                break

    def _sift_down(self, index: int) -> None:
        size = len(self._data)
        while True:
            smallest = index
            left = 2 * index + 1
            right = 2 * index + 2

            if left < size and self._data[left] < self._data[smallest]:
                smallest = left
            if right < size and self._data[right] < self._data[smallest]:
                smallest = right

            if smallest != index:
                self._data[index], self._data[smallest] = self._data[smallest], self._data[index]
                index = smallest
            else:
                break

    def __repr__(self) -> str:
        return f"MinHeap({self._data})"


# ═══════════════════════════════════════════════════════════════════════════
# Union-Find — Disjoint set with path compression and union by rank
# ═══════════════════════════════════════════════════════════════════════════

class UnionFind:
    """Disjoint set — find and union with path compression."""

    def __init__(self, n: int) -> None:
        self.parent = list(range(n))
        self.rank = [0] * n
        self.components = n

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # Path compression
        return self.parent[x]

    def union(self, x: int, y: int) -> bool:
        root_x, root_y = self.find(x), self.find(y)
        if root_x == root_y:
            return False

        if self.rank[root_x] < self.rank[root_y]:
            self.parent[root_x] = root_y
        elif self.rank[root_x] > self.rank[root_y]:
            self.parent[root_y] = root_x
        else:
            self.parent[root_y] = root_x
            self.rank[root_x] += 1

        self.components -= 1
        return True

    def connected(self, x: int, y: int) -> bool:
        return self.find(x) == self.find(y)

    def __repr__(self) -> str:
        return f"UnionFind(components={self.components})"


# ═══════════════════════════════════════════════════════════════════════════
# Trie — Prefix tree for string operations
# ═══════════════════════════════════════════════════════════════════════════

class Trie:
    """Prefix tree for efficient string prefix operations."""

    class _Node:
        def __init__(self) -> None:
            self.children: dict[str, "Trie._Node"] = {}
            self.is_end = False

    def __init__(self) -> None:
        self.root = Trie._Node()

    def insert(self, word: str) -> None:
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = Trie._Node()
            node = node.children[char]
        node.is_end = True

    def search(self, word: str) -> bool:
        node = self._find_node(word)
        return node is not None and node.is_end

    def starts_with(self, prefix: str) -> bool:
        return self._find_node(prefix) is not None

    def _find_node(self, prefix: str) -> Optional["Trie._Node"]:
        node = self.root
        for char in prefix:
            if char not in node.children:
                return None
            node = node.children[char]
        return node

    def autocomplete(self, prefix: str) -> list[str]:
        node = self._find_node(prefix)
        if node is None:
            return []
        results: list[str] = []
        self._collect_words(node, prefix, results)
        return results

    def _collect_words(self, node: "Trie._Node", current: str, results: list[str]) -> None:
        if node.is_end:
            results.append(current)
        for char, child in sorted(node.children.items()):
            self._collect_words(child, current + char, results)

    def __repr__(self) -> str:
        return f"Trie(size={self.root is not None})"


# ═══════════════════════════════════════════════════════════════════════════
# LRU Cache — Least Recently Used eviction
# ═══════════════════════════════════════════════════════════════════════════

class LRUCache(Generic[K, V]):
    """LRU cache with O(1) get/put using OrderedDict-like behavior."""

    def __init__(self, capacity: int) -> None:
        self._capacity = capacity
        self._cache: dict[K, V] = {}
        self._order: deque[K] = deque()

    def get(self, key: K) -> Optional[V]:
        if key not in self._cache:
            return None
        self._order.remove(key)
        self._order.append(key)
        return self._cache[key]

    def put(self, key: K, value: V) -> None:
        if key in self._cache:
            self._order.remove(key)
        elif len(self._cache) >= self._capacity:
            oldest = self._order.popleft()
            del self._cache[oldest]
        self._cache[key] = value
        self._order.append(key)

    def size(self) -> int:
        return len(self._cache)

    def __repr__(self) -> str:
        return f"LRUCache({dict(self._cache)})"


# ═══════════════════════════════════════════════════════════════════════════
# Demo / Test
# ═══════════════════════════════════════════════════════════════════════════

def main() -> None:
    print("=== Ruva Data Structures Library (Python) ===\n")

    # ArrayList test
    arr: ArrayList[int] = ArrayList()
    arr.add(10)
    arr.add(20)
    arr.add(30)
    print(f"ArrayList: {arr.get(0)}, {arr.get(1)}, {arr.get(2)}")
    print(f"  size: {arr.size()}")

    # Stack test
    stack: Stack[int] = Stack()
    stack.push(100)
    stack.push(200)
    stack.push(300)
    print(f"\nStack pop: {stack.pop()}, {stack.pop()}")

    # HashMap test
    hmap: HashMap[str, int] = HashMap()
    hmap.put("one", 1)
    hmap.put("two", 2)
    hmap.put("three", 3)
    print(f"\nHashMap: one={hmap.get('one')}, two={hmap.get('two')}")

    # BinarySearchTree test
    tree: BinarySearchTree[int] = BinarySearchTree()
    for val in [5, 3, 7, 1, 4, 6, 8]:
        tree.insert(val)
    print(f"\nBST inorder: {tree.inorder()}")
    print(f"  height: {tree.height()}")

    # MinHeap test
    heap: MinHeap[int] = MinHeap()
    for val in [5, 3, 7, 1, 4, 6, 8]:
        heap.push(val)
    order = []
    while not heap.is_empty():
        order.append(heap.pop())
    print(f"\nMinHeap poll order: {order}")

    # Union-Find test
    uf = UnionFind(6)
    uf.union(0, 1)
    uf.union(1, 2)
    uf.union(3, 4)
    print(f"\nUnionFind: 0~2 connected? {uf.connected(0, 2)}")
    print(f"  0~3 connected? {uf.connected(0, 3)}")
    print(f"  components: {uf.components}")

    # Trie test
    trie = Trie()
    for word in ["apple", "app", "application", "bat", "ball", "band"]:
        trie.insert(word)
    print(f"\nTrie search('apple'): {trie.search('apple')}")
    print(f"  starts_with('app'): {trie.starts_with('app')}")
    print(f"  autocomplete('app'): {trie.autocomplete('app')}")
    print(f"  autocomplete('ba'): {trie.autocomplete('ba')}")

    # LRU Cache test
    cache: LRUCache[int, str] = LRUCache(3)
    cache.put(1, "one")
    cache.put(2, "two")
    cache.put(3, "three")
    cache.get(1)  # Makes 1 most recent
    cache.put(4, "four")  # Evicts 2
    print(f"\nLRU Cache: get(1)={cache.get(1)}, get(2)={cache.get(2)}, get(4)={cache.get(4)}")

    print("\nAll data structures functional.")


if __name__ == "__main__":
    main()
